### DISCUSSION TITLE
Deleting images from Github Package Registry

### DISCUSSION THREAD

**Discussion by kachow6** [community/community#21609] [New to GitHub]
<p>As part of my CI Actions workflow, I push a docker image of the latest branch commit to GPR and pull it onto a remote docker-machine for testing. The image is tagged with the commit sha like so to prevent concurrent workflows from mixing:</p>
<pre><code>docker.pkg.github.com/$GITHUB_REPOSITORY/$REPOSITORY_NAME:$GITHUB_SHA
</code></pre>
<p>After my tests are complete, I would like to delete/remove the image from Github Package Registry but I can’t seem to find the functionality after reviewing the docker x gpr documentation here: <a href="https://help.github.com/en/github/managing-packages-with-github-packages/configuring-docker-for-use-with-github-packages" rel="nofollow noopener">https://help.github.com/en/github/managing-packages-with-github-packages/configuring-docker-for-use-with-github-packages</a> or package management documentation here: <a href="https://help.github.com/en/github/managing-packages-with-github-packages/deleting-a-package" rel="nofollow noopener">https://help.github.com/en/github/managing-packages-with-github-packages/deleting-a-package</a>.</p>
<p>Is there a way to delete/remove docker images from GPR? If so can someone point me in the right direction or does this functionality not exist yet?</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hi <a class="mention" href="/kachow6">@kachow6</a>,</p>
<p>Thank you for being here! This functionality is currently in the works, for now you might message us via  <a href="https://github.com/contact" rel="nofollow noopener">https://github.com/contact</a> with the info, for our support folks to take care of it.</p>

**Comment by kachow6** [comment]
<p>Hi thanks for the response <a class="mention" href="/andreagriffiths11">@andreagriffiths11</a>,</p>
<p>Where can I find the update/release log for Github Package Registry? I would like to be notified when this feature is released.</p>

**Comment by phramz** [comment]
<p>Hi <a class="mention" href="/andreagriffiths11">@andreagriffiths11</a> ! Any update on this topic yet? </p>
<p>Best regards, Max</p>

**Comment by jhelbig** [comment]
<p>Just wanted to chime in here.  I want to use the image registry for my team since it will speed up our development efforts having the builds so close to the source, but not being able to manage the images with full CRUD really diminishes our desire to use it.</p>
<p>Without this abililty I’m going to have to make a choice soon to use a service like ECR to manage our images.  This doesn’t make much sense since we’re already paying for an enterprise account.  Once we make a choice to use an external service across all of our applications, migrating back to GHR won’t make much sense because of the manhours it would take to migrate everything back to yet another image registry.</p>
<p>I’m not asking for an explanation as to why it’s not here, but I can’t wait forever for this functionality and I assume there are a couple thousand other teams that are in the same situation.  That seems like a fairly heavy loss of untapped revenue, enough that perhaps this could have a priority escalation or at the very least far more communcation about the status of the improvement.</p>
<p>Thanks,</p>
<p>Jake</p>

**Comment by mmuth** [comment]
<p>We also have the same issues. We would like to delete images and we’d be glad if that was supported soon.</p>

**Comment by arshabh-finanta** [comment]
<p><a href="https://help.github.com/en/packages/publishing-and-managing-packages/deleting-a-package" rel="nofollow noopener">https://help.github.com/en/packages/publishing-and-managing-packages/deleting-a-package</a></p>

**Comment by kachow6** [comment]
<p>Since initially making this post I was able to come up with a solution using curl, jq, and the github graphql api that worked for my specific use case (deleting a gpr image based on the version name within an actions workflow). Sharing it here in hopes that it helps someone else. Note, I had to create a github personal access token with read, write, and delete privileges for GPR packages, which i stored as a repository secret.</p>
<p>Get the 10 most recent package versions and save the version id with name that matches the current commit sha:</p>
<pre><code>- name: get gpr package version id from first 10 packages
      run: |
        echo "::set-env name=VERSION_ID::$(curl -X POST \
        -H "Accept: application/vnd.github.package-deletes-preview+json" \
        -H "Authorization: Bearer ${{ secrets.GPR_PERSONAL_ACCESS_TOKEN }}" \
        -d "{\"query\":\"query {repository(owner:\\\""$YOUR_ORG_NAME"\\\", name:\\\""$REPOSITORY_NAME"\\\") {registryPackages(first:10) {edges{node{id, name, versions(first:10){edges {node {id, version}}}}}}}}\"}" \
        https://api.github.com/graphql | jq '.data.repository.registryPackages.edges[].node.versions.edges[].node' | jq -s '.' | jq --arg GITHUB_SHA $GITHUB_SHA '.[] | select(.version==$GITHUB_SHA) | .id')"
</code></pre>
<p>Attempt to delete the version based on $VERSION_ID:</p>
<pre><code>- name: delete gpr package version
      id: attempt_delete
      run: |
        echo "::set-output name=delete::$(curl -X POST \
        -H "Accept: application/vnd.github.package-deletes-preview+json" \
        -H "Authorization: Bearer ${{ secrets.GPR_PERSONAL_ACCESS_TOKEN }}" \
        -d "{\"query\":\"mutation { deletePackageVersion(input:{packageVersionId:\\\"${{ env.VERSION_ID }}\\\"}) { success }}\"}" \
        https://api.github.com/graphql)"
</code></pre>
<p>Check result of deletion step and fail workflow if deletion failed (otherwise passes regardless):</p>
<pre><code>- name: check if delete was successful
      run: |
        if [-n $(echo "$DELETE_RESULT" | jq '.data.deletePackageVersion')]
          then
          echo "$DELETE_RESULT" | jq '.errors' | jq '.[].message'
          exit 1
        fi
      env:
        DELETE_RESULT: ${{ steps.attempt_delete.outputs.delete }}
</code></pre>

**Comment by mattwebbio** [comment]
<p>Thanks so much for this. I tag my Docker images with the branch name - for anyone doing something similiar, I was able to modify the first query to get the version ID like this:</p>
<pre><code>- name: Get Docker image's GPR ID
        run: |
          echo "::set-env name=VERSION_ID::$(curl -X POST \
          -H "Accept: application/vnd.github.packages-preview+json" \
          -H "Authorization: Bearer ${{ github.token }}" \
          -d "{\"query\":\"query {repository(owner:\\\""&lt;org name&gt;"\\\", name:\\\""&lt;repo name&gt;"\\\") {packages(names:[\\\""&lt;package name&gt;"\\\"], first:1) {edges{node{id, name, version(version:\\\""${{ github.event.pull_request.head.ref }}"\\\"){id, version}}}}}}\"}" \
          https://api.github.com/graphql | jq '.data.repository.packages.edges[].node.version.id')"
</code></pre>

**Comment by clarkbw** [comment, marked as answer]
<p>If you’re looking for a simpler delete solution we maintain an action that helps you automate this process <a href="https://github.com/actions/delete-package-versions" rel="nofollow noopener">https://github.com/actions/delete-package-versions</a></p>

**Comment by kachow6** [comment]
<p>Updating this to be the recommended solution thanks</p>

**Comment by kachow6** [comment]
<p>Updated script to get first 10 packages after registryPackages was deprecated on 2020-04-01:</p>
<pre><code class="lang-auto">- name: get gpr package version id from first 10 packages
      run: |
        echo "::set-env name=VERSION_ID::$(curl -X POST \
        -H "Accept: application/vnd.github.package-deletes-preview+json" \
        -H "Authorization: Bearer ${{ secrets.GPR_PERSONAL_ACCESS_TOKEN }}" \
        -d "{\"query\":\"query {repository(owner:\\\""$YOUR_ORG_NAME"\\\", name:\\\""$REPOSITORY_NAME"\\\") {packages(first:10) {edges{node{id, name, versions(first:10){edges {node {id, version}}}}}}}}\"}" \
        https://api.github.com/graphql | jq '.data.repository.packages.edges[].node.versions.edges[].node' 
</code></pre>

**Comment by mr-linch** [comment]
<p>if you need delete images by tag you can checkout our action</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/bots-house/ghcr-delete-image-action" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/3163427/181065821-9da384b0-8cda-49fa-a6ec-1985812fd15f.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/b/f/bf946fa91639d32e3b2613002113ce3823d5472f_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/b/f/bf946fa91639d32e3b2613002113ce3823d5472f_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/b/f/bf946fa91639d32e3b2613002113ce3823d5472f.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/b/f/bf946fa91639d32e3b2613002113ce3823d5472f_2_10x10.png"></div>

<h3><a href="https://github.com/bots-house/ghcr-delete-image-action" target="_blank" rel="noopener nofollow ugc">bots-house/ghcr-delete-image-action</a></h3>


  <p><span class="label1">♻️ Delete GitHub Container Registry image by tag. Contribute to bots-house/ghcr-delete-image-action development by creating an account on GitHub.</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
